Turn on compile with NVSHMEM

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #154568
* __->__ #154538

Before:
`USE_NVSHMEM=1` need to be explicit set in build environment.

After:
`USE_NVSHMEM=1` is the default for CUDA/Rocm on Linux.

kwen2501 committed:
Turn on compile with NVSHMEM

[ghstack-poisoned]

Skylion007 commented:
We'll need https://pypi.nvidia.com/nvidia-nvshmem-cu12/ added to the release builds at some point if they aren't already

kwen2501 commented:
Thanks @Skylion007 . Thinking the same. Do you know how I can do that?

Skylion007 commented:
At the very least, you need to add it to here: https://github.com/pytorch/pytorch/blob/24980d2641318d70ddc23dc9ff9b2f5f70245c4d/.github/scripts/generate_binary_build_matrix.py#L74 and have @atalman or someone with write access upload a binary to the S3 so the nightly builds do not fail (you need to regenerate the YAMLs after editing this file)

kwen2501 commented:
I added the extra dependency in https://github.com/pytorch/pytorch/pull/154568.

kwen2501 committed:
Update on "Turn on compile with NVSHMEM"


Before:
`USE_NVSHMEM=1` need to be explicit set in build environment.

After:
`USE_NVSHMEM=1` is the default for CUDA/Rocm on Linux. 

[ghstack-poisoned]

kwen2501 commented:
@ngimel  Yes, at compile time, NVSHMEM is auto detected by tools/setup_helpers/cmake.py. (I am going to refactor it into findNVSHMEM.cmake).

At user side, I added https://github.com/pytorch/pytorch/pull/154568 to make NVSHMEM an install dependency.

kwen2501 committed:
Update on "Turn on compile with NVSHMEM"


Before:
`USE_NVSHMEM=1` need to be explicit set in build environment.

After:
`USE_NVSHMEM=1` is the default for CUDA/Rocm on Linux. 

[ghstack-poisoned]
